In [1]:
with open("agh.scss", "w", encoding="utf8") as f:
    f.write(r"""
/*-- scss:defaults --*/
$body-bg: #FFFFFF;
$body-color: #111111;
$link-color: #1D65A6;

$presentation-heading-color: #0B3C5D;
$presentation-font-size-root: 32px;
$presentation-heading-font: "Times New Roman", serif;
$font-family-sans-serif: "Times New Roman", serif;
$font-family-serif: "Times New Roman", serif;

/*-- scss:rules --*/
.reveal {
  font-family: "Times New Roman", serif;
  color: #111111;
}

.reveal h1,
.reveal h2,
.reveal h3 {
  color: #0B3C5D;
  font-family: "Times New Roman", serif;
}

.reveal strong {
  color: #1D65A6;
}

.reveal a {
  color: #1D65A6;
}

.slide-logo {
  right: 10px !important;
  max-height: 10% !important;
  width: auto !important;
}

figure {
  text-align: center;
}

figure img {
  border-radius: 5px;
}

div.MathJax_Display,
.reveal .math.display {
  color: #0B3C5D;
}

.reveal table {
  font-size: 0.85em;
}

.reveal th {
  color: #0B3C5D;
}

.reveal td {
  color: #111111;
}
""")

In [2]:
with open("refs.bib", "w", encoding="utf8") as f:
    f.write(r"""
@book{Feynman_1963,
  year = {1963},
  title = {The Feynman Lectures on Physics},
  author = {Feynman, R. P.},
  publisher = {Addison-Wesley}
}
""")

In [3]:
# enforcing inline vector graphics
%config InlineBackend.figure_format = 'svg'

# trigger font manager cache rebuild
from matplotlib import figure

figure.Figure()

# AGH colours + colour-blind-safe linestyles + font sizes
import matplotlib
from cycler import cycler

import matplotlib.pyplot as plt
import matplotlib as mpl

mpl.rcParams.update({
    # ===== Font =====
    "font.family": "serif",
    "font.serif": ["Times New Roman"],
    
    # ===== Text sizes =====
    "font.size": 18,
    "axes.titlesize": 24,
    "axes.labelsize": 20,
    "xtick.labelsize": 16,
    "ytick.labelsize": 16,
    "legend.fontsize": 16,
    
    # ===== Figure =====
    "figure.figsize": (12, 5),
    "figure.facecolor": "white",
    "axes.facecolor": "white",
    
    # ===== Text / axes colors =====
    "text.color": "black",
    "axes.labelcolor": "black",
    "axes.edgecolor": "#222222",
    "xtick.color": "#222222",
    "ytick.color": "#222222",
    
    # ===== Grid =====
    "axes.grid": True,
    "grid.color": "#D9E2F2",
    "grid.linestyle": "--",
    "grid.linewidth": 0.8,
    
    # ===== Lines =====
    "lines.linewidth": 2.2,
    "lines.markersize": 7,
    
    # ===== Color palette =====
    "axes.prop_cycle": cycler(color=[
        "#0B3C5D",  # deep navy
        "#328CC1",  # medium blue
        "#5DA9E9",  # light blue
        "#1D65A6",  # steel blue
        "#7FB3D5"   # pale blue
    ]),
    
    # ===== Legend =====
    "legend.frameon": False,
    "legend.handlelength": 1.4,
    
    # ===== Export quality =====
    "figure.dpi": 150,
})

## Project goal

Create a clear, beginner-friendly section showing how to build and tune tables in **Quarto + Python**.

The section will focus on:

- text labels
- numerical values
- time values
- left, center, and right alignment
- number formatting
- significant digits
- clean table output for reports and slides

## Why this section matters

Tables are common in scientific and technical work, but raw outputs are often hard to read.

A good table should make the data structure obvious:

| Type of value | Typical formatting problem |
|:--|:--|
| Text labels | inconsistent alignment |
| Numbers | too many decimals |
| Small / large values | need scientific notation or significant digits |
| Time values | unreadable timestamps |
| Units | need clear column labels |

## Proposed contribution

I will prepare a practical tutorial section titled:

> **Handling and formatting tables in Quarto with Python**

The section will include:

1. Basic Markdown tables.
2. Tables generated from Python.
3. Alignment rules.
4. Formatting numerical values.
5. Formatting time values.
6. A final “raw table vs polished table” example.

## Scope of the tutorial

The examples will use one small dataset inspired by measurement data.

The table will contain:

| Column | Meaning |
|:--|:--|
| `sample` | text label |
| `mass_kg` | number |
| `time_s` | number |
| `measured_at` | time value |
| `uncertainty` | small number needing scientific notation |

In [4]:
import pandas as pd

raw = pd.DataFrame({
    "sample": ["A", "B", "C", "D"],
    "mass_kg": [1.23456, 12.0, 0.004321, 1234.567],
    "time_s": [0.0123, 15.678, 1234.5, 89.001],
    "measured_at": pd.to_datetime([
        "2026-05-14 09:15",
        "2026-05-14 09:45",
        "2026-05-14 10:30",
        "2026-05-14 11:05",
    ]),
    "uncertainty": [0.0001234, 0.04567, 12.345, 0.000009876],
})

raw

,sample,mass_kg,time_s,measured_at,uncertainty
0,A,1.234560,0.0123,2026-05-14 09:15:00,0.000123
1,B,12.000000,15.6780,2026-05-14 09:45:00,0.045670
2,C,0.004321,1234.5000,2026-05-14 10:30:00,12.345000
3,D,1234.567000,89.0010,2026-05-14 11:05:00,0.000010


## Part 1: Markdown table alignment

Quarto supports normal Markdown tables.

Alignment is controlled with colons in the separator row:

```markdown
| Label | Value | Unit |
|:------|------:|:----:|
| mass  | 1.23  | kg   |
| time  | 12.5  | s    |
```

Result:

| Label | Value | Unit |
|:------|------:|:----:|
| mass  | 1.23  | kg   |
| time  | 12.5  | s    |

## Part 2: Generate tables from Python

A `pandas.DataFrame` can be displayed directly in Quarto.

This is useful when the table is produced by calculations rather than typed manually.

In [5]:
raw[["sample", "mass_kg", "time_s"]]

,sample,mass_kg,time_s
0,A,1.234560,0.0123
1,B,12.000000,15.6780
2,C,0.004321,1234.5000
3,D,1234.567000,89.0010


## Part 3: Number formatting

The tutorial will explain common Python format specifiers:

| Specifier | Meaning | Example |
|:--|:--|:--|
| `.2f` | 2 digits after decimal point | `12.35` |
| `.3g` | 3 significant digits | `0.00432` |
| `.2e` | scientific notation | `1.23e-04` |
| `,.2f` | thousands separator | `1,234.57` |

In [6]:
formatted = raw.copy()

formatted["mass_kg"] = formatted["mass_kg"].map(lambda x: f"{x:.3g}")
formatted["time_s"] = formatted["time_s"].map(lambda x: f"{x:.2f}")
formatted["uncertainty"] = formatted["uncertainty"].map(lambda x: f"{x:.2e}")
formatted["measured_at"] = formatted["measured_at"].dt.strftime("%H:%M")

formatted

,sample,mass_kg,time_s,measured_at,uncertainty
0,A,1.23,0.01,09:15,1.23e-04
1,B,12,15.68,09:45,4.57e-02
2,C,0.00432,1234.50,10:30,1.23e+01
3,D,1.23e+03,89.00,11:05,9.88e-06


## Part 4: Significant digits

A key point will be the difference between decimal places and significant digits.

```python
f"{x:.3f}"  # 3 decimal places
f"{x:.3g}"  # 3 significant digits
```

This matters for measurements, where values may have very different magnitudes.

In [7]:
values = [1234.567, 12.34567, 0.1234567, 0.0001234567]

pd.DataFrame({
    "raw value": values,
    "3 decimal places": [f"{x:.3f}" for x in values],
    "3 significant digits": [f"{x:.3g}" for x in values],
})

,raw value,3 decimal places,3 significant digits
0,1234.567000,1234.567,1.23e+03
1,12.345670,12.346,12.3
2,0.123457,0.123,0.123
3,0.000123,0.000,0.000123


## Part 5: Time values

Time data should usually be converted from full timestamps to a format that matches the purpose of the table.

Examples:

| Format | Meaning |
|:--|:--|
| `%H:%M` | hour and minute |
| `%Y-%m-%d` | date |
| `%Y-%m-%d %H:%M` | date and time |

In [8]:
time_demo = raw[["sample", "measured_at"]].copy()

time_demo["time only"] = time_demo["measured_at"].dt.strftime("%H:%M")
time_demo["date only"] = time_demo["measured_at"].dt.strftime("%Y-%m-%d")
time_demo["date and time"] = time_demo["measured_at"].dt.strftime("%Y-%m-%d %H:%M")

time_demo

,sample,measured_at,time only,date only,date and time
0,A,2026-05-14 09:15:00,09:15,2026-05-14,2026-05-14 09:15
1,B,2026-05-14 09:45:00,09:45,2026-05-14,2026-05-14 09:45
2,C,2026-05-14 10:30:00,10:30,2026-05-14,2026-05-14 10:30
3,D,2026-05-14 11:05:00,11:05,2026-05-14,2026-05-14 11:05


## Part 6: Polished table with labels

The final example will rename columns so that the table reads like a publication table.

In [9]:
final_table = formatted.rename(columns={
    "sample": "Sample",
    "mass_kg": "Mass / kg",
    "time_s": "Time / s",
    "measured_at": "Measured at",
    "uncertainty": "Uncertainty",
})

final_table

,Sample,Mass / kg,Time / s,Measured at,Uncertainty
0,A,1.23,0.01,09:15,1.23e-04
1,B,12,15.68,09:45,4.57e-02
2,C,0.00432,1234.50,10:30,1.23e+01
3,D,1.23e+03,89.00,11:05,9.88e-06


## Optional: styled HTML table

For slides or HTML output, pandas styling can tune alignment visually.

The tutorial can show this as an optional advanced example.

In [10]:
styled_table = (
    final_table.style
    .hide(axis="index")
    .set_properties(subset=["Sample"], **{"text-align": "left"})
    .set_properties(subset=["Mass / kg", "Time / s", "Uncertainty"], **{"text-align": "right"})
    .set_properties(subset=["Measured at"], **{"text-align": "center"})
)

styled_table

Sample,Mass / kg,Time / s,Measured at,Uncertainty
A,1.23,0.01,09:15,1.23e-04
B,12,15.68,09:45,4.57e-02
C,0.00432,1234.50,10:30,1.23e+01
D,1.23e+03,89.00,11:05,9.88e-06
